Issue/2304

mehul0810 opened the pull request:
## Description
This PR resolves #2304 and resolves #2023 

## How Has This Been Tested?
I've manually tested this feature with multiple scenarios:
- Email Access Enabled and accessing the donation receipt via donation success email. 
- Email Access Enabled and accessing donation history using donation history page directly.
- Email Access Disabled and accessing the donation history page directly will ask for login.

Also, I've tried above scenario's in incognito windows as well.

## Types of changes
Bug fix (non-breaking change which fixes an issue)
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

DevinWalker commented on includes/misc-functions.php:
```diff
@@ -1499,3 +1535,47 @@ function give_get_core_settings_json( $file_name ) {
+				<div class="give-security-column give-security-description-wrap">
+					<?php
+					echo sprintf(
+						__( 'For security reason, please confirm your email address (%s) to view your complete donation history.', 'give' ),
```
Typo here. Change to:

> For security reasons, 

Missing "s" on reasons

DevinWalker commented on includes/class-give-email-access.php:
```diff
@@ -202,27 +209,24 @@ public function send_email( $customer_id, $email ) {
 
-		$message = __( 'You or someone in your organization requested an access link be sent to this email address. This is a temporary access link for you to view your donation information. Click on the link below to view:', 'give' ) . "\n\n";
-		$message .= '<a href="' . esc_url( $access_url ) . '" target="_blank">' . __( 'Access Donation Details &raquo;', 'give' ) . '</a>' . "\n\n";
+		$message = sprintf( __( 'Please click the link to access your donation history on %s. If you did not request this email, please contact %s.', 'give' ), get_bloginfo( 'url' ), get_bloginfo( 'admin_email' ) ) . "\n\n";
```
Hyperlink `get_bloginfo( 'url' )` - right now it's not linked

DevinWalker commented on includes/class-give-email-access.php:
```diff
@@ -202,27 +209,24 @@ public function send_email( $customer_id, $email ) {
-		$message = __( 'You or someone in your organization requested an access link be sent to this email address. This is a temporary access link for you to view your donation information. Click on the link below to view:', 'give' ) . "\n\n";
-		$message .= '<a href="' . esc_url( $access_url ) . '" target="_blank">' . __( 'Access Donation Details &raquo;', 'give' ) . '</a>' . "\n\n";
+		$message = sprintf( __( 'Please click the link to access your donation history on %s. If you did not request this email, please contact %s.', 'give' ), get_bloginfo( 'url' ), get_bloginfo( 'admin_email' ) ) . "\n\n";
+		$message .= '<a href="' . esc_url( $access_url ) . '" target="_blank">' . __( 'Click here to view donation history &raquo;', 'give' ) . '</a>' . "\n\n";
```
Change to just "View your donation history &raquo;"

mehul0810 commented:
@DevinWalker I've implemented the suggested changes. Also, tested the functionality again.

DevinWalker commented on includes/admin/settings/class-settings-general.php:
```diff
@@ -75,23 +75,47 @@ public function get_settings() {
 						),
+						array(
+							'id'      => 'limit_display_donations',
+							'name'    => __( 'Limit Display of Donations', 'give' ),
```
Change text to:

> Limit Donations Displayed

DevinWalker commented on includes/admin/settings/class-settings-general.php:
```diff
@@ -75,23 +75,47 @@ public function get_settings() {
+						array(
+							'id'      => 'limit_display_donations',
+							'name'    => __( 'Limit Display of Donations', 'give' ),
+							'desc'    => __( 'Number of donations to display when non logged-in user access donation history page with purchase session deactivated.', 'give' ),
```
Change text to:

> Adjusts the number of donations displayed to a non logged-in user when they attempt to access the donation history page without an active session. For security reasons, it's best to leave this at 1-3 donations.

DevinWalker commented on includes/admin/settings/class-settings-general.php:
```diff
@@ -75,23 +75,47 @@ public function get_settings() {
+							'css'     => 'width:12em;',
+						),
+						array(
+							'id'      => 'limit_email_throttle_time',
```
Remove this option and just provide filters, it's too confusing.

DevinWalker commented on includes/admin/settings/class-settings-general.php:
```diff
@@ -75,23 +75,47 @@ public function get_settings() {
+							'type'    => 'number',
+							'css'     => 'width:12em;',
+						),
+						array(
```
Remove this option and just provide filter also

mehul0810 commented:
@DevinWalker I've implemented required changes

DevinWalker commented:
## Reviewing Findings Round 2

### Issue 1: If _only_ a single donation has been made by a donor then don't display email option:

![2017-11-29_18-09-49](https://user-images.githubusercontent.com/1571635/33409506-94c3ffea-d530-11e7-91bb-2d17436c37d5.gif)

### Issue 2: Full donation history doesn't appear to be loading properly

See video here: https://goo.gl/RjdfKG

mehul0810 commented:
@DevinWalker Please review it. I've updated the code as per the suggestions and tested it.

DevinWalker commented:
## Review Findings

Please watch video @mehul0810 https://drive.google.com/open?id=14XNRoawdvoKdMZ1BfVdo_nxUOZONuz3N

We're getting closer.

mehul0810 commented:
@DevinWalker I've updated PR as per the video attached as well as discussion over call.
- Error Message Text change to `Please refresh the page in 5 minutes before requesting a new donation history access link.` (Also, number of minutes will change based on the filter amount set.)
- Remove extra period at the end of the Error Message.
- Fixed the default to 5 mins which was mistakenly set to 5 hours.
- Hide the `Confirm Email` button block and just show message when locked out to improve UX.

Please review and let me know if there are any further changes.

ravinderk commented on includes/shortcodes.php:
```diff
@@ -44,8 +44,11 @@ function give_donation_history( $atts ) {
 
+		// Display donation history link only if it is not accessed via Receipt Access Link.
+		if ( ! give_get_receipt_session() ) {
+			echo '<a href="' . esc_url( give_get_history_page_uri() ) . '">&laquo; ' . __( 'Return to All Donations', 'give' ) . '</a>';
```
@mehul0810 use `sprintf`

mehul0810 replied:
@ravinderk Fixed!

ravinderk commented on templates/email-login-form.php:
```diff
@@ -6,10 +6,8 @@
-$email     = isset( $_POST['give_email'] ) ? $_POST['give_email'] : '';
 
-// Declare Variables.
+$email            = isset( $_POST['give_email'] ) ? $_POST['give_email'] : '';
```
@mehul0810 use `give_clean`

mehul0810 replied:
@ravinderk Fixed!

raftaar1191 commented:
@mehul0810 
A small error message is shown when user log-in with the link
Video: http://screencast-o-matic.com/watch/cblhXM2gGE

<details>
```
### WordPress Environment ###

Home URL: http://give17.local
Site URL: http://give17.local
WP Version: 4.9.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
Permalink Structure: /%postname%/
Show on Front: posts
Table Prefix Length: 3
Table Prefix Status: Acceptable
Admin AJAX: Inaccessible
Registered Post Statuses: publish, future, draft, pending, private, trash, auto-draft, inherit, refunded, failed, revoked, cancelled, abandoned, processing, preapproval

### Server Environment ###

Hosting Provider: DBH: localhost, SRV: give17.local
TLS Connection: Connection uses TLS 1.2
TLS Connection: Probably Okay
Server Info: nginx/1.10.1
PHP Version: 7.0.3
PHP Post Max Size: 1,000 MB
PHP Time Limit: 900
PHP Max Input Vars: 3000
PHP Max Upload Size: 1,000 MB
cURL Version: ❌ 7.38.0, OpenSSL/1.0.1t - We recommend a minimum cURL version of 7.40.
SUHOSIN Installed: –
MySQL Version: ❌ 5.5.55 - We recommend a minimum MySQL version of 5.6. See: WordPress Requirements
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
gzip: ✔
GD Graphics Library: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Give Configuration ###

Give Version: 1.8.17
Upgraded From: 1.8.17
Test Mode: Enabled
Currency Code: USD
Currency Position: Before
Decimal Separator: .
Thousands Separator: ,
Success Page: http://give17.local/donation-confirmation/
Failure Page: http://give17.local/donation-failed/
Donation History Page: http://give17.local/donation-history/
Give Forms Slug: /donations/
Enabled Payment Gateways: Test Donation, Offline Donation
Default Payment Gateway: Test Donation
PayPal IPN Verification: Enabled
PayPal IPN Notifications: N/A
Admin Email Notifications: Enabled
Donor Email Access: Enabled

### Session Configuration ###

Give Use Sessions: Enabled
Session: Disabled

### Active Give Add-ons ###


### Other Active Plugins ###


### Inactive Plugins ###

Give - Fee Recovery: by WordImpress – 1.3.3
Give - PayPal Pro Gateway: by WordImpress – 1.1.4

### Theme ###

Name: Twenty Sixteen
Version: 1.4
Author URL: https://wordpress.org/
Child Theme: No – If you're modifying Give on a parent theme you didn't build personally, then we recommend using a child theme. See: How to Create a Child Theme

```
</details>

raftaar1191 commented:
Getting PHP Notices when login from the link:

```[05-Dec-2017 11:59:51 UTC] PHP Notice:  Undefined index: post_type in /app/public/wp-content/plugins/Give/includes/payments/class-payments-query.php on line 484
[05-Dec-2017 11:59:51 UTC] PHP Stack trace:
[05-Dec-2017 11:59:51 UTC] PHP   1. {main}() /app/public/index.php:0
[05-Dec-2017 11:59:51 UTC] PHP   2. require() /app/public/index.php:17
[05-Dec-2017 11:59:51 UTC] PHP   3. require_once() /app/public/wp-blog-header.php:19
[05-Dec-2017 11:59:51 UTC] PHP   4. include() /app/public/wp-includes/template-loader.php:74
[05-Dec-2017 11:59:51 UTC] PHP   5. get_sidebar() /app/public/wp-content/themes/twentysixteen/page.php:40
[05-Dec-2017 11:59:51 UTC] PHP   6. locate_template() /app/public/wp-includes/general-template.php:110
[05-Dec-2017 11:59:51 UTC] PHP   7. load_template() /app/public/wp-includes/template.php:647
[05-Dec-2017 11:59:51 UTC] PHP   8. require_once() /app/public/wp-includes/template.php:688
[05-Dec-2017 11:59:51 UTC] PHP   9. dynamic_sidebar() /app/public/wp-content/themes/twentysixteen/sidebar.php:13
[05-Dec-2017 11:59:51 UTC] PHP  10. WP_Widget->display_callback() /app/public/wp-includes/widgets.php:742
[05-Dec-2017 11:59:51 UTC] PHP  11. WP_Widget_Recent_Posts->widget() /app/public/wp-includes/class-wp-widget.php:372
[05-Dec-2017 11:59:51 UTC] PHP  12. WP_Query->__construct() /app/public/wp-includes/widgets/class-wp-widget-recent-posts.php:71
[05-Dec-2017 11:59:51 UTC] PHP  13. WP_Query->query() /app/public/wp-includes/class-wp-query.php:3328
[05-Dec-2017 11:59:51 UTC] PHP  14. WP_Query->get_posts() /app/public/wp-includes/class-wp-query.php:3222
[05-Dec-2017 11:59:51 UTC] PHP  15. apply_filters_ref_array() /app/public/wp-includes/class-wp-query.php:2539
[05-Dec-2017 11:59:51 UTC] PHP  16. WP_Hook->apply_filters() /app/public/wp-includes/plugin.php:244
[05-Dec-2017 11:59:51 UTC] PHP  17. Give_Payments_Query->custom_orderby() /app/public/wp-includes/class-wp-hook.php:286
[05-Dec-2017 11:59:51 UTC] PHP Notice:  Undefined index: post_type in /app/public/wp-content/plugins/Give/includes/payments/class-payments-query.php on line 484
[05-Dec-2017 11:59:51 UTC] PHP Stack trace:
[05-Dec-2017 11:59:51 UTC] PHP   1. {main}() /app/public/index.php:0
[05-Dec-2017 11:59:51 UTC] PHP   2. require() /app/public/index.php:17
[05-Dec-2017 11:59:51 UTC] PHP   3. require_once() /app/public/wp-blog-header.php:19
[05-Dec-2017 11:59:51 UTC] PHP   4. include() /app/public/wp-includes/template-loader.php:74
[05-Dec-2017 11:59:51 UTC] PHP   5. get_sidebar() /app/public/wp-content/themes/twentysixteen/page.php:40
[05-Dec-2017 11:59:51 UTC] PHP   6. locate_template() /app/public/wp-includes/general-template.php:110
[05-Dec-2017 11:59:51 UTC] PHP   7. load_template() /app/public/wp-includes/template.php:647
[05-Dec-2017 11:59:51 UTC] PHP   8. require_once() /app/public/wp-includes/template.php:688
[05-Dec-2017 11:59:51 UTC] PHP   9. dynamic_sidebar() /app/public/wp-content/themes/twentysixteen/sidebar.php:13
[05-Dec-2017 11:59:51 UTC] PHP  10. WP_Widget->display_callback() /app/public/wp-includes/widgets.php:742
[05-Dec-2017 11:59:51 UTC] PHP  11. WP_Widget_Recent_Posts->widget() /app/public/wp-includes/class-wp-widget.php:372
[05-Dec-2017 11:59:51 UTC] PHP  12. WP_Query->__construct() /app/public/wp-includes/widgets/class-wp-widget-recent-posts.php:71
[05-Dec-2017 11:59:51 UTC] PHP  13. WP_Query->query() /app/public/wp-includes/class-wp-query.php:3328
[05-Dec-2017 11:59:51 UTC] PHP  14. WP_Query->get_posts() /app/public/wp-includes/class-wp-query.php:3222
[05-Dec-2017 11:59:51 UTC] PHP  15. apply_filters_ref_array() /app/public/wp-includes/class-wp-query.php:2539
[05-Dec-2017 11:59:51 UTC] PHP  16. WP_Hook->apply_filters() /app/public/wp-includes/plugin.php:244
[05-Dec-2017 11:59:51 UTC] PHP  17. Give_Payments_Query->custom_orderby() /app/public/wp-includes/class-wp-hook.php:286

```

raftaar1191 commented:
For https://github.com/WordImpress/Give/pull/2395#issuecomment-349283468
-> I had a check in Firefox in Normal Window and still able to regenerated the issues at my end

mehul0810 commented:
@raftaar1191 I've fixed the issue presented in your video attached which is due to token getting expired earlier. However, I'm unable to reproduce the php notices issue mentioned. Can you please confirm that you can still reproduce the php notice issue. Let me know your thoughts and a quick call for discussion, if required.

raftaar1191 commented:
Thanks, @mehul0810  for fixing it. I had a check it and it's working absolutely fine. Great Work!!


@mehul0810 I have able to regenerate the notices but it not related to the above issues. So I am creating new issues for this

DevinWalker commented:
## Another Review

The email access submission form is not working and needs some UI help.

Please watch video: https://goo.gl/B6hbuJ

#### Additional Move this Option

This option is relevant whether email access is enabled or not so it shouldn't be under the conditional show/hide fields:

![2017-12-06_14-03-45](https://user-images.githubusercontent.com/1571635/33687900-83c0819c-da8e-11e7-8d5c-d2801e0c52c1.png)

Please move it out.

mehul0810 commented:
@DevinWalker I've created a testing video as i'm unable to reproduce the issue. Also, I've created video with Email Access enabled scenario which is having issues. 

**Video:** https://drive.google.com/open?id=1A1gk4TVaXxhHZd8Ce83Dif5IA119vC4v

Please check the video and let me know your thoughts. Also, let me know if quick call is required to review the PR together.

raftaar1191 commented:
As review with @mehul0810 over the call, I am not also not able to reproduce the issues

DevinWalker commented:
@mehul0810 I'm merging this because it's pretty much working and we need to get 1.8.17 out the door but I'm going to open an issue to address a few additional items I'd like to see implemented.
